### ISSUE TITLE
NullReferenceException inside Pedigree Viewer

### ISSUE THREAD

**Issue by phroph** [cadon/ARKStatsExtractor#1056]
This is a bit of a strange issue I'm running into. I can summarize it as follows:

I started using this tool midway into a game, after already doing a few generations of Snow Owl and Rex breeding. However, before I started using the tool, some of my breeders had died without having their data exported.

When I import from my ARK savefile, it recognizes the Snow Owl/Rex parents inside the Pedigree list and lists them as unavailable. However, no entries are entered into Library for me to post-humously add in stats (for at least one of the dead breeders I have the exact stats and ARK ID for the dino). When I try to click the pedigree entries for descendants of the missing dinos, the app crashes with a NPE:

`
************** Exception Text **************
System.NullReferenceException: Object reference not set to an instance of an object.
   at ARKBreedingStats.CreatureColored.getColoredCreature(Int32[] colorIds, Species species, Boolean[] enabledColorRegions, Int32 size, Int32 pieSize, Boolean onlyColors, Boolean dontCache)
   at ARKBreedingStats.PedigreeCreature.set_Creature(Creature value)
   at ARKBreedingStats.PedigreeCreature..ctor(Creature creature, Boolean[] enabledColorRegions, Int32 comboId)
   at ARKBreedingStats.Pedigree.createParentsChild(Creature creature, Int32 x, Int32 y, Boolean drawWithNoParents, Boolean highlightCreature)
   at ARKBreedingStats.Pedigree.createPedigree()
   at ARKBreedingStats.Pedigree.setCreature(Creature centralCreature, Boolean forceUpdate)
   at ARKBreedingStats.Pedigree.listViewCreatures_SelectedIndexChanged(Object sender, EventArgs e)
   at System.Windows.Forms.ListView.OnSelectedIndexChanged(EventArgs e)
   at System.Windows.Forms.ListView.WmReflectNotify(Message& m)
   at System.Windows.Forms.ListView.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
`

This is a painful issue because probably half of our Snow Owls/Rexs at least descend from the missing breeders.

I have tried manually adding the missing breeder using the collected stats/ARK ID. This creates a dino with the correct stats in the library, however the dinos pedigrees are still linked to the unavailable dino instead of the backfilled dino data. I can manually change the genealogy to fix this however next time I import the ARK save file the dinos are all linking back to the unavailable dino and are ignoring the one I manually added.

**TL;DR**, the problem boils down to:
-Unavailable entries in the pedigree cannot have entries directly created for them.
-Manually set geneology information gets wiped over when loading an ARK save, to the point of replacing N/A with a broken link (or a working link with the broken link). Every time after reloading an ARK save I have to go through all the dinos and remove the missing links. This is tedious. The alternative is usually Dino Export for everything but I find this glitchy and harder to deal with in general.
-When ARK Stats Extractor tries to access a pedigree for a dino with a missing entry in the link, it throws a NPE rather than gracefully handling the null and stopping there.

If any of these gets fixed it would unblock my use case and make this tool a lot easier to work with. It's more frustrating in this case because I have some of the exact stats to backfill this data and have no way to seamlessly integrate this. At the very least I can click "Continue" on the JIT exception and continue my life, but it does some weird things to the UI.

**Comment by cadon** [comment]
Thanks for the report. I'll fix that in the next release.
The reason for the exception is that imported creatures save the id of their parents, even if these are not yet available in the library. That way, once the parents are imported to that library, they're linked correctly to their offspring. In ASB these missing parents are created as placeholder creatures with the according id set, and these placeholders are not visible in the library (they serve no purpose there), but only in the pedigree. Each time a library is loaded, missing placeholder creatures are created, so if the user imports them, they can be connected correctly in their pedigree (so deleting them manually does only work until you load the library again).
The placeholders can only be replaced by a real creature if the according creature with the id is imported to ASB. It's not possible to add these creatures manually (ok, one way to do it would be to create an export file with the correct ID and import that), because ASB only trusts IDs that come directly from the game or export files, to ensure consistency.

If you have no way to get the export files of the missing parents, you can select a creature and change its parents manually. Anyway, the exception will be fixed soon. Sorry for the inconveniance.

**Comment by phroph** [comment]
Thanks for the quick response! That makes sense to me, if I have some time I may try to reconstruct the ark exports manually, though at least getting rid of the exception is a good QoL booster!

Cheers
